### PR TITLE
Support indices contiguity per batch and non-contiguous values in sparse compressed tensors

### PR DESCRIPTION
**Pull request by pearu**
Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91331
* #90849
* __->__ #91243

cc @nikitaved @cpuhrsch @amjames @bhosmer

### CODE REVIEW

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h [hunk 1]
```diff
@@ -1,17 +1,19 @@
 #pragma once
-
+#include <ATen/Dispatch.h>
 #include <ATen/Tensor.h>
+#include <ATen/Utils.h>
 #include <ATen/native/TensorIterator.h>
-#include <ATen/Dispatch.h>
 #include <ATen/native/sparse/Macros.h>
+#include <iostream>
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:7]: to be removed

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h [hunk 2]
```diff
@@ -241,52 +266,79 @@ void _validate_compressed_sparse_indices_kernel(
 
     const auto batch_dims = cidx.sizes().slice(0, cidx.dim() - 1);
     const auto batch_count = indexCount(batch_dims);
-    const auto batch_idx = at::arange(batch_count, cidx.options()).view(batch_dims).unsqueeze_(-1);
+    const auto batch_idx =
+        at::arange(batch_count, cidx.options()).view(batch_dims).unsqueeze_(-1);
+
+    const Tensor batch_shape = at::tensor(batch_dims, idx.options());
+    const Tensor batch_idx_stride =
+        at::tensor(idx.strides().slice(0, idx.dim() - 1), idx.options());
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:274]: Shape and strides are int64. Let us probably keep it that way for these newly created tensors.

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h [hunk 3]
```diff
@@ -241,52 +266,79 @@ void _validate_compressed_sparse_indices_kernel(
...
+        all_idx_offsets.reshape({-1, 1})
+            .div(batch_idx_stride, "floor") // TODO: support zero strides
+            .remainder(batch_shape)
+            .mul(batch_idx_stride)
+            .sum(1)
+            .eq(all_idx_offsets)
+            .nonzero()
+            .to(idx.dtype())
+            .view(batch_dims)
+            .unsqueeze_(-1);
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:288]: These ops could be done on the CPU, and then moved to the GPU if needed.

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp [hunk 4]
```diff
@@ -174,6 +184,15 @@ void _validate_sparse_compressed_tensor_args_worker(const Tensor& compressed_ind
               dense_ndim >= 0,
               "values must have dimensionality > sum of batch and block dimensionalities (=",
               batch_ndim, " + ", block_ndim, ") but got ", values.dim());
+
+  // 3.5
+  TORCH_CHECK(plain_indices.stride(plain_indices.dim() - 1) == 1,
+              "expected ", plain_indices_name, " to be a contiguous tensor per batch");
+
```
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/native/sparse/SparseCsrTensor.cpp:191]: Will use `plain_indices.stride(-1)`, thanks for the hint!

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h (hunk 3 above)
Comment (nikitaved) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:288]: Maybe, if that is of use, there is a function `make_offset_calculator` that allows to map a linear index to a multi-dim offset.

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h (hunk 1 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:7]: fixed

File: aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp [hunk 5]
```diff
@@ -864,10 +864,10 @@ void add_out_dense_sparse_csr_cpu(
     return;
   }
 
-  auto valuesBuffer = src_values.to(commonDtype).view({-1, src_values.size(-1)});
+  auto valuesBuffer = src_values.to(commonDtype).reshape({-1, src_values.size(-1)});
   resultBuffer = resultBuffer.view({-1, out.size(-2), out.size(-1)});
-  auto src_crow_indices = src.crow_indices().view({-1, src.crow_indices().size(-1)});
-  auto src_col_indices = src.col_indices().view({-1, src.col_indices().size(-1)});
+  auto src_crow_indices = src.crow_indices().reshape({-1, src.crow_indices().size(-1)});
+  auto src_col_indices = src.col_indices().reshape({-1, src.col_indices().size(-1)});
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:870]: The CPP version uses just `reshape` while the CUDA version is doing `reshape + contiguous`. Either way, if `reshape` is called (or `flatten`), the batch dimensions become regular, and there is no need in sophisticated maps from a linear index to a multi-dim offset.
It seems like dropping `reshape` is best for performance, and we could use `make_offset_calculator` to keep the CPU and GPU kernel relatively intact. Otherwise it is best to re-implement these kernels and use `TensorIterator.for_each`.

Comment (pearu) [reply aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:870]: > The CPP version uses just reshape while the CUDA version is doing reshape + contiguous

This is due to CPU version using accessor indexing while the CUDA version uses pointer indexing.

The rest of the comments are OT to this PR but I agree that the CPU and GPU kernels could be unified here.

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:870]: Why are they different? They use the same code...

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:870]: Sorry, @pearu, I thought these are inside invariant checks, but they turn out to be in the `add` op. Apologies for the noise...

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h [hunk 6]
```diff
@@ -241,52 +266,79 @@ void _validate_compressed_sparse_indices_kernel(
...
+    const Tensor batch_shape = at::tensor(batch_dims, idx.options());
+    const Tensor batch_idx_stride =
+        at::tensor(idx.strides().slice(0, idx.dim() - 1), idx.options());
+    const Tensor all_idx_offsets = at::arange(
+        batch_idx_stride.mul(batch_shape.sub(1)).sum().add(1).item(),
+        idx.options());
+    const Tensor idx_offsets =
+        all_idx_offsets.reshape({-1, 1})
+            .div(batch_idx_stride, "floor") // TODO: support zero strides
+            .remainder(batch_shape)
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:281]: Is it easy to fix? Do we get an error message and not a segfault if there are zero strides?

Comment (pearu) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:281]: `make_offset_calculator` appears to be CUDA specific but there seems to be analogous code for the CPU. I'll investigate that..

Re zero strides: I think we should support it because otherwise, test_bmm needs to use an unorthodox way of constructing CSR tensors. I suggest leaving it as a follow-up PR within this ghstack as in addition to generalizing the above, the generate_simple_inputs need to be updated accordingly and it is unpredictable how other existing tests behave on zero-strided inputs.

Comment (nikitaved) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:281]: Then, I guess, it would be great to just error out, and then remove it in a follow-up as to avoid any potential segfaults...

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h [hunk 7]
```diff
@@ -241,52 +272,83 @@ void _validate_compressed_sparse_indices_kernel(
...
     const auto batch_count = indexCount(batch_dims);
-    const auto batch_idx = at::arange(batch_count, cidx.options()).view(batch_dims).unsqueeze_(-1);
+    const auto batch_idx =
+        at::arange(batch_count, cidx.options()).view(batch_dims).unsqueeze_(-1);
+
+    const auto idx_ndims = idx.dim();
+    const auto idx_sizes = idx.sizes();
+    const auto idx_strides = idx.strides();
+    const auto idx_contiguous_strides = c10::contiguous_strides(idx_sizes);
+    std::array<int64_t, 3 * MAX_DIMS> idx_sizes_and_strides;
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:282]: nit: having 3 separate arrays might be a bit easier to understand.

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h [hunk 8]
```diff
@@ -241,52 +272,83 @@ void _validate_compressed_sparse_indices_kernel(
...
+                const int64_t* sizes = idx_sizes_and_strides.data();
+                const int64_t* strides = sizes + idx_ndims;
+                const int64_t* contiguous_strides = sizes + 2 * idx_ndims;
+                int64_t idx_offset = 0;
+                // assuming idx contiguity per batch:
+                const int64_t batch_idx_ = batch_idx * sizes[idx_ndims - 1];
+                for (int i = 0; i < idx_ndims - 1; i++) {
+                  idx_offset +=
+                      ((batch_idx_ / contiguous_strides[i]) % sizes[i]) *
+                      strides[i];
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:334]: Do we really need `% sizes[i]`?

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h (hunk 7 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h]: Fixed.

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h (hunk 6 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h]: Fixed, the latest commit supports zero strides.

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h (hunk 3 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h]: The computation of idx offsets is moved to the kernel.

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h (hunk 2 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h]: Resolved.

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h [hunk 9]
```diff
@@ -241,52 +272,75 @@ void _validate_compressed_sparse_indices_kernel(
 
     const auto batch_dims = cidx.sizes().slice(0, cidx.dim() - 1);
     const auto batch_count = indexCount(batch_dims);
-    const auto batch_idx = at::arange(batch_count, cidx.options()).view(batch_dims).unsqueeze_(-1);
+    const auto batch_idx =
+        at::arange(batch_count, cidx.options()).view(batch_dims).unsqueeze_(-1);
+
+    const auto idx_ndims = idx.dim();
+    std::array<int64_t, MAX_DIMS> idx_sizes, idx_strides;
+    for (int i = 0; i < idx_ndims; i++) {
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:280]: We should probably error out if `idx_ndims > MAX_DIMS`, otherwise it is a UB.

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h [hunk 10]
```diff
@@ -241,52 +272,75 @@ void _validate_compressed_sparse_indices_kernel(
...
+                // NOTE: the implementation below is sync-less, but,
+                // unfortunately, work is not guaranteed to be well-balanced
+                // between different threads.
+                int64_t idx_offset = 0;
+                // assuming idx contiguity per batch:
+                int64_t tmp = batch_idx * idx_sizes[idx_ndims - 1];
+                for (int i = idx_ndims - 1; i >= 0; i--) {
+                  idx_offset += (tmp % idx_sizes[i]) * idx_strides[i];
+                  tmp /= idx_sizes[i];
+                }
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:327]: Division and mod are slow on CUDA, especially compared to all other ops in the kernel. I am fine with that for now, but it would not hurt to optimize this case for contiguous inputs.

Comment (pearu) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:327]: I can eliminate `mod` here but overall this feels like an over-optimization considering that the invariants validation can be disabled if maximum performance is required.

Comment (pearu) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h]: Eliminated `mod`.

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h (hunk 9 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h]: Fixed: sizes and strides are now passed to the kernel via `Tensor` instance and we don't depend on `MAX_DIMS` anymore.

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h [hunk 11]
```diff
@@ -276,11 +269,9 @@ void _validate_compressed_sparse_indices_kernel(
...
 
     const auto idx_ndims = idx.dim();
-    std::array<int64_t, MAX_DIMS> idx_sizes, idx_strides;
-    for (int i = 0; i < idx_ndims; i++) {
-      idx_sizes[i] = idx.size(i);
-      idx_strides[i] = idx.stride(i);
-    }
+
+    Tensor idx_sizes_ = at::tensor(idx.sizes(), idx.options().dtype(kLong));
+    Tensor idx_strides_ = at::tensor(idx.strides(), idx.options().dtype(kLong));
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:274]: minor nit: we can create a buffer tensor on the cpu, copy sizes and strides into it, move it to CUDA, then `unbind` it into two separate CUDA tensors for only a single data transfer. But maybe it is an over-optimization...

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h [hunk 12]
```diff
@@ -322,8 +316,9 @@ void _validate_compressed_sparse_indices_kernel(
                 // assuming idx contiguity per batch:
                 int64_t tmp = batch_idx * idx_sizes[idx_ndims - 1];
                 for (int i = idx_ndims - 1; i >= 0; i--) {
-                  idx_offset += (tmp % idx_sizes[i]) * idx_strides[i];
-                  tmp /= idx_sizes[i];
+                  int64_t div = tmp / idx_sizes[i];
+                  idx_offset += (tmp - div * idx_sizes[i]) * idx_strides[i];
+                  tmp = div;
                 }
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:322]: 👍

File: aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h (hunk 11 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h]: Done.

File: torch/testing/_internal/common_utils.py [hunk 13]
```diff
@@ -2709,6 +2714,21 @@ def generate_values(base, densesize):
...
+        def non_contiguous_copy(t, dim=-1, offset=0):
+            # return a copy of t that is non-contiguous along the
+            # given dimension and with the given storage offset
+            self.assertTrue(t.is_contiguous())
+            if dim < 0:
+                dim = dim + t.ndim
+            assert dim >= 0 and dim < t.ndim
+            step = max(2, offset + 1)
+            tmp = torch.zeros((*t.shape[:dim], t.shape[dim] * step, *t.shape[dim + 1:]), dtype=t.dtype, device=t.device)
+            tmp[(*((slice(None),) * dim), slice(offset, None, step))].copy_(t)
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2726]: nit: `tmp[..., slice(...)]` should work as well.

**Review by nikitaved — APPROVED** [review]
LGTM. Thank you, @pearu !

File: torch/testing/_internal/common_utils.py (hunk 13 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2726]: Thanks! I'll apply this in some follow-up.

**Comment by pearu** [comment]
@YXIE14, @842974287, @terrychenism, @fuqianz, or @Xirider , could you approve this PR? Thanks!

File: torch/testing/_internal/common_utils.py (hunk 13 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py]: Using `Ellipsis` does not work because `offset::step` is not the last dimension.

**Review by amjames — APPROVED** [review]

**Review by lezcano — APPROVED** [review]
